Introduce loadable pseudo-random number generators.

Don-Ward opened the pull request:
N.B. The PRNG interface introduced here is subject to review and may change.

PRNGs are per-thread and may be changed with the new loadrng() standard
function. Two PRNGs are present:
   IconEx   A reimplementation of the traditional generator
            as a loadable library.
   Rabbit   A cryptographically strong PRNG from the eSTREAM project.
Other PRNGs -- probably RanLux and the Mersenne Twister -- will follow.

etherx-dev reviewed (approved):
My head hurts... but i know that stumbling upon this project just made it worthwhile lol.

Don-Ward committed:
Update build.yml to add a rnglib test build.

Add a test for --enable-rnglibs

Don-Ward commented:
After initial feedback, it was decided to split up the PR into several smaller commits

Don-Ward committed:
Preparation for loadable Random Number Generators (rnglibs).

Add data structures and definitions.

Don-Ward committed:
Add rnglib <-> runtime API and new standard functions.

Add runtime routines callable by a rnglib:
   getRngState
   getInitialBits
   putErrorCode
Add rnglib routines callable by the runtime:
   startRng
   putSeed
   getRandomFpt
   getErrorText
   getRandomBits  -- optional
   getRandomInt   -- optional

Add new standard functions:
   loadrng
   rngbits
   rngbitstring
   rngval

Don-Ward committed:
Add rnglib definitions to &features and the Unicon preprocessor

Add "loadable RNGs" to &features.
Add _RNG_LIBRARY to the preprocessor predefined symbols.

Don-Ward committed:
Modify &random and the ? operator to use rnglibs.

Don-Ward committed:
Add prnglib failure codes and boilerplate error messages.

Don-Ward committed:
Make GC aware of (and collect) rnglib structures.

Don-Ward committed:
Add the rngIconEx loadable library.

RngIconEx is a reimplemtation of the (Icon) built-in generator
as a loadable rnglib. It serves as an example for how implement a
loadable rng library.

Don-Ward committed:
Add the rngRbt loadable library.

Rabbit is a cryptographically secure random number generator from the
ESTREAM project. rngRbt implements it as a loadable rng library.

Don-Ward committed:
Add the rngRlx++ loadable library.

rngRlx++ is a reimplementation in C of the RanLux++ generator.

Don-Ward committed:
Add the rngGSL loadable library.

rngGSL isn't a random number generator as such; it provides an interface
to the generators in the GNU Scientific Library, if that is installed.

Don-Ward committed:
Modify the Unicon build process to include rng loadable libraries.

Add a new argument --enable-rnglibs to configure.
Detect the presence (or otherwise) of the GNU Scientific Library.

Don-Ward committed:
Add rnglib test programs.

Don-Ward committed:
Document the new rnglib features in Appendix G of the Unicon book.

Don-Ward committed:
Initial draft of a utr that describes the rnglib changes.

Don-Ward committed:
Fix typo in the name of the rngRbt directory.

Don-Ward committed:
runtime: avoid name collision with getrandom() from libc [Jafar Al-Gharaibeh]

This is a slight rework of commit 27b385c3. It is needed because the
PRNG code uses unicon_getrandom() in fmisc.r as well as init.r

Don-Ward committed:
Accomodate differences between 32-bit and 64-bit builds of rngRbt

Don-Ward committed:
GC: Replace hand calulated offsets with compiler calculations.

The hand calculated values in the firstd and firstp arrays are replaced with
compiler calculations using offsetof(). This should make future changes easier
and less prone to error.

Jafaral committed:
runtime: classify passthrough markers

Replace runtime passthru sites with explicit raw C markers.
Leave parsed C constructs visible to RTT where the translator now
understands them.

Signed-off-by: Jafar Al-Gharaibeh <to.jafar@gmail.com>

Jafaral committed:
rtt: regenerate bison output for host and raw C passthrough

Update rttgram.y and regenerate rttparse.c and ltoken.h.

Signed-off-by: Jafar Al-Gharaibeh <to.jafar@gmail.com>

Jafaral committed:
rtt: add host and raw C passthrough

Teach RTT to parse the runtime C constructs that are safe
to understand directly. Keep explicit output directives for
code that must still bypass RTT parsing or preprocessing.

Signed-off-by: Jafar Al-Gharaibeh <to.jafar@gmail.com>

Jafaral committed:
rtt: give thread-local specifiers their own token

Map __thread and _Thread_local to Thread_local instead of Static so
storage-class analysis (e.g. just_type) cannot confuse TLS with static.
Regenerate the Bison parser tables.

Signed-off-by: Jafar Al-Gharaibeh <to.jafar@gmail.com>

Jafaral committed:
rtt: document pinned %token numbers and Bison-generated parser policy

Signed-off-by: Jafar Al-Gharaibeh <to.jafar@gmail.com>

Jafaral committed:
rtt: regenerate yyerror.c for Bison parser; document merr and YYUNDEF

Regenerate yyerror.c with merr -B so errtab state keys match the current
automaton; size yymaxstate/errtab to YYNSTATES - 1 / YYNSTATES from rttparse.c.
Include stdlib.h for calloc.

rttgram.y: note that token code 368 is reserved as YYUNDEF.

Signed-off-by: Jafar Al-Gharaibeh <to.jafar@gmail.com>

Don-Ward commented:
@greptile review

Don-Ward commented:
@greptile review

Don-Ward commented:
@greptile Please do a full review as if you are seeing this pull request for the first time. Include all issues except the acknowledged data race in puterrorcode (which is very, very, unlikely to be an issue in practice because the designers of new rnglibs will cooperate so as to not interfere with each other).

Don-Ward commented:
@greptile review

Don-Ward committed:
test: Remove configuration specific values from mutex test

Some configurations have a different number of mutexes, which leads to
a test failure (because the test output is not the same as the expected
result). This commit changes the output so that exact values are no longer
part of the output. The actual tests are unchanged.

Don-Ward commented:
@greptile review

Don-Ward commented:
@greptile review

Don-Ward commented:
@greptile review

Don-Ward commented:
@greptile review

Don-Ward commented on src/runtime/fmisc.r:
```diff
@@ -2506,6 +2506,565 @@ end
 
+#ifdef RngLibrary
+
+void *getRngState( void)
+{
+  CURTSTATE();
+
+  if (curtstate->rng != NULL) {
+    if (curtstate->Kywd_ran.vword.bptr != NULL) {
+      return &(curtstate->Kywd_ran.vword.bptr->List.listhead->Intarray.a[1]);
+    }
+  }
+  return NULL;
+}
```
@greptile This analysis is incorrect. Even if a GC is triggered in another thread whilst the `putSeed` routine is executing, the GC cannot proceed until all the other threads (including the thread running 'putSeed') are paused pending the GC. Because that will only happen after the `putSeed` routine has finished, the postulated movement of the GC-managed block cannot happen while the routine is operating.

Don-Ward commented:
@greptile review

Don-Ward commented:
@greptile In a previous review it was stated "rmisc.r returns D_Intarray as the descriptor type for &random when a PRNG library is loaded, but every other site in this PR that builds that descriptor uses D_List. Code that inspects the type tag will see an inconsistent view of the variable." and in the latest review it was stated "src/runtime/rmisc.r (wrong dword type tag),".  Please supply all instances where the code returns D_List as part of the desciptor value for &random. Give filenames and line numbers.

Don-Ward commented:
@greptile review.  Also, in a previous review it was stated "rmisc.r returns D_Intarray as the descriptor type for &random when a PRNG library is loaded, but every other site in this PR that builds that descriptor uses D_List. Code that inspects the type tag will see an inconsistent view of the variable." and in the latest review it was stated "src/runtime/rmisc.r (wrong dword type tag),". Please supply all instances where the code returns D_List as part of the desciptor value for &random. Give filenames and line numbers.

Don-Ward commented:
@greptile review
